spec/consensus/light-client.md

```diff
@@ -257,28 +267,49 @@ Towards Lite Client Completeness:
 
 
 ```go
-func Bisection(h1,h2,trustlevel) bool{
-  if CheckSupport(h1,h2,trustlevel) {
-    return true
-  }
-  if h2.Header.height == h1.Header.height + 1 {
-    // we have adjacent headers that are not matching (failed
-    // the CheckSupport)
-    // we could submit evidence here
-    return false
-  }
-  pivot := (h1.Header.height + h2.Header.height) / 2
-  hp := Commit(pivot)
-    // ask a full node for header of height pivot
-  Store(hp)  
-    // store header hp locally
-  if Bisection(h1,hp,trustlevel) {
-    // only check right branch if hp is trusted
-    // (otherwise a lot of unnecessary computation may be done)
-    return Bisection(hp,h2,trustlevel)
-  }
-  else {
-    return false
+// return (true, nil) in case we can trust header h2 based on header h1; otherwise return (false, error) where error captures the nature of the error.
+func Bisection(h1,h2,trustlevel) (bool, error) {
+  assume h1.Header.Height < h2.header.Height
+
+  ok, err = CheckSupport(h1,h2,trustlevel)
+  if (ok or err != nil) return (ok, err)
+
+  // we cannot verify h2 based on h1, so we try to move trusted header closer to h2 so we can verify h2
+  th := h1 // th is trusted header
+  while th.Header.Height <= h2.Header.height - 1 do {
+       // try to move trusted header forward with stored headers
+       ih := th
+       for all stored headers h s.t ih.Header.Height < h.Header.height < h2.Header.height do {  // try to move trusted header forward
+            // we assume here that iteration is done in the order of header heights
+            ok, err = CheckSupport(th,h,trustlevel)
+            if err != nil { return (ok, err) }
+            if ok {
+                th = h
+            }
+        }
+
+        // at this point we have potentially updated th based on stored headers so we try to verify h2 based on new trusted header
+        ok, err = CheckSupport(th,h2,trustlevel)
+        if (ok or err != nil) return (ok, err)
+
+        // we cannot verify h2 based on th, so we try to move trusted header closer to h2 by downloading header(s) between th and h2
+        endHeight = h2.Header.height
```

Comment:
endHeight obscures the real value / does not bring anything to the table actually. I'd recommend just using h2.Header.Height directly